src/middleware.rs

```diff
@@ -1,51 +1,69 @@
+    pub pool: Pool<PostgresConnectionManager>,
 }
 
 impl PostgresMiddleware {
-    pub fn new(connect_str: &str,
-               ssl_mode: SslMode,
-               num_connections: u32,
-               error_handler: Box<HandleError<::r2d2_postgres::Error>>)
-                    -> Result<PostgresMiddleware, Box<StdError>> {
-        let manager = try!(PostgresConnectionManager::new(connect_str, ssl_mode));
-
-        let config = Config::builder()
-          .pool_size(num_connections)
-          .error_handler(error_handler)
-          .build();
+    /// Create middleware using defaults
+    ///
+    /// The middleware will be setup with no ssl and the r2d2 defaults.
+    pub fn new(db_url: &str) -> Result<PostgresMiddleware, Box<Error>> {
+        let manager = try!(PostgresConnectionManager::new(db_url, SslMode::None));
+        let pool = try!(Pool::new(Config::default(), manager));
 
-        let pool = try!(Pool::new(config, manager));
+        Ok(PostgresMiddleware { pool: pool })
+    }
 
-        Ok(PostgresMiddleware { pool: Arc::new(pool) })
+    /// Create middleware using pre-built `r2d2::Pool`
+    ///
+    /// This allows the caller to create and configure the pool with specific settings.
+    pub fn with_pool(pool: Pool<PostgresConnectionManager>) -> PostgresMiddleware {
+        PostgresMiddleware { pool: pool }
     }
 }
 
-impl Key for PostgresMiddleware { type Value = Arc<Pool<PostgresConnectionManager>>; }
+impl Key for PostgresMiddleware { type Value = Pool<PostgresConnectionManager>; }
 
 impl<D> Middleware<D> for PostgresMiddleware {
     fn invoke<'mw, 'conn>(&self, req: &mut Request<'mw, 'conn, D>, res: Response<'mw, D>) -> MiddlewareResult<'mw, D> {
         req.extensions_mut().insert::<PostgresMiddleware>(self.pool.clone());
+
         Ok(Continue(res))
     }
 }
 
+/// Add `db_conn()` helper method to `nickel::Request`
+///
+/// This trait must only be used in conjunction with `PostgresMiddleware`.
+///
+/// On error, the method returns a tuple per Nickel convention. This allows the route to use the
+/// `try_with!` macro.
+///
+/// Example:
+///
+/// ```ignore
+/// app.get("/my_counter", middleware! { |request, response|
+/// 	let db = try_with!(response, request.pg_conn());
+/// });
+/// ```
 pub trait PostgresRequestExtensions {
-    fn db_conn(&self) -> PooledConnection<PostgresConnectionManager>;
+    fn pg_conn(&self) -> Result<PooledConnection<PostgresConnectionManager>, (StatusCode, GetTimeout)>;
 }
 
 impl<'a, 'b, D> PostgresRequestExtensions for Request<'a, 'b, D> {
-    fn db_conn(&self) -> PooledConnection<PostgresConnectionManager> {
-        self.extensions().get::<PostgresMiddleware>().unwrap().get().unwrap()
+    fn pg_conn(&self) -> Result<PooledConnection<PostgresConnectionManager>, (StatusCode, GetTimeout)> {
+        self.extensions()
+            .get::<PostgresMiddleware>()
+            .expect("PostgresMiddleware must be registered with Nickel::utilize()")
```

Comment:
This runtime failure might want to be more explicit that it could be an ordering bug, i.e. that using `db_conn` in a handler registered before registering the middleware is an error.